pandas/core/ops/array_ops.py

```diff
@@ -132,3 +153,233 @@ def na_arithmetic_op(left, right, op, str_rep, eval_kwargs):
+
+        if res_values is NotImplemented:
+            res_values = invalid_comparison(lvalues, rvalues, op)
+        if is_scalar(res_values):
+            raise TypeError(
+                "Could not compare {typ} type with Series".format(typ=type(rvalues))
+            )
+
+    return res_values
+
+
+def na_logical_op(x, y, op):
+    try:
+        result = op(x, y)
+    except TypeError:
+        if isinstance(y, np.ndarray):
+            # bool-bool dtype operations should be OK, should not get here
+            assert not (is_bool_dtype(x.dtype) and is_bool_dtype(y.dtype))
+            x = ensure_object(x)
+            y = ensure_object(y)
+            result = libops.vec_binop(x, y, op)
+        else:
+            # let null fall thru
+            assert lib.is_scalar(y)
+            if not isna(y):
+                y = bool(y)
+            try:
+                result = libops.scalar_binop(x, y, op)
+            except (
+                TypeError,
+                ValueError,
+                AttributeError,
+                OverflowError,
+                NotImplementedError,
+            ):
+                raise TypeError(
+                    "cannot compare a dtyped [{dtype}] array "
+                    "with a scalar of type [{typ}]".format(
+                        dtype=x.dtype, typ=type(y).__name__
+                    )
+                )
+
+    return result
+
+
+def logical_op(
+    left: Union[np.ndarray, ABCExtensionArray], right: Any, op
+) -> Union[np.ndarray, ABCExtensionArray]:
+    """
+    Evaluate a logical operation `|`, `&`, or `^`.
+
+    Parameters
+    ----------
+    left : np.ndarray or ExtensionArray
+    right : object
+        Cannot be a DataFrame, Series, or Index.
+    op : {operator.and_, operator.or_, operator.xor}
+        Or one of the reversed variants from roperator.
+
+    Returns
+    -------
+    ndarrray or ExtensionArray
+    """
+    from pandas.core.ops import should_extension_dispatch, dispatch_to_extension_op
+
+    fill_int = lambda x: x
+
+    def fill_bool(x, left=None):
+        # if `left` is specifically not-boolean, we do not cast to bool
+        if x.dtype.kind in ["c", "f", "O"]:
+            # dtypes that can hold NA
+            mask = isna(x)
+            if mask.any():
+                x = x.astype(object)
+                x[mask] = False
+
+        if left is None or is_bool_dtype(left.dtype):
+            x = x.astype(bool)
+        return x
+
+    is_self_int_dtype = is_integer_dtype(left.dtype)
+
+    right = lib.item_from_zerodim(right)
+    if is_list_like(right) and not hasattr(right, "dtype"):
+        # e.g. list, tuple
+        right = construct_1d_object_array_from_listlike(right)
+
+    # NB: We assume extract_array has already been called on left and right
+    lvalues = left
+    rvalues = right
+
+    if should_extension_dispatch(lvalues, rvalues):
+        res_values = dispatch_to_extension_op(op, lvalues, rvalues)
+
+    else:
```

Comment:
can be direct else/if

Reply:
Not without duplicating L378-380 i think